packages/gatsby/src/utils/webpack.config.js

```diff
@@ -339,11 +339,21 @@ module.exports = async (
         // get schema to pass to eslint config and program for directory
         const { schema, program } = store.getState()
 
+        const isCustomEslint = hasLocalEslint(program.directory)
+
         // if no local eslint config, then add gatsby config
-        if (!hasLocalEslint(program.directory)) {
+        if (!isCustomEslint) {
           configRules = configRules.concat([rules.eslint(schema)])
         }
 
+        // Enforce fast-refresh rules even with local eslint config
+        if (
+          isCustomEslint &&
+          process.env.GATSBY_HOT_LOADER === `fast-refresh`
+        ) {
+          configRules = configRules.concat([rules.eslintRequired()])
+        }
```

Comment:
Basically the main piece. We add another minimal loader only when there is user-defined eslint config. When there is no custom config - we use a single default loader (with fast-refresh rules added to it)